openssl/src/crypto/signature.rs

```diff
@@ -0,0 +1,291 @@
+//!
+//! // Generate public/private keypairs
+//! let mut keypair = PKey::new();
+//! let mut privkey = PKey::new();
+//! let mut pubkey = PKey::new();
+//! keypair.gen(1024);
+//! privkey.load_priv(&keypair.save_priv()[..]);
+//! pubkey.load_pub(&keypair.save_pub()[..]);
+//!
+//! // generate some data
+//! let data1: Vec<u8> = (0..25).cycle().take(1024).collect();
+//! let data2: Vec<u8> = (100..150).cycle().take(1024 * 5).collect();
+//!
+//! // sign some content with private key
+//! let mut signer = Signer::new(hash::Type::SHA512, &privkey);
+//! signer.write_all(&data1[..]).unwrap();
+//! signer.write_all(&data2[..]).unwrap();
+//! let signature = signer.finish();
+//!
+//! // verify content using signature and public key
+//! let mut verifier = Verifier::new(hash::Type::SHA512, &pubkey);
+//! verifier.write_all(&data1[..]).unwrap();
+//! verifier.write_all(&data2[..]).unwrap();
+//! assert!(verifier.finish(&signature[..]));
+//! ```
+
+// pkey not used in structure but we need to ensure that the PKey
+// lives as long as the structure
+#![allow(dead_code)]
+
+use std::ptr;
+use std::io::{self, Write};
+use libc;
+
+use ::ffi;
+use ::crypto::pkey::PKey;
+use ::crypto::hash;
+
+/// Signature Generation Context
+pub struct Signer<'a> {
+    ctx: *mut ffi::EVP_MD_CTX,
+    pkey: &'a PKey,
+}
+
+/// Signature Verification Context
+pub struct Verifier<'a> {
+    ctx: *mut ffi::EVP_MD_CTX,
+    pkey: &'a PKey,
+}
+
+macro_rules! openssl_assert {
+    ($expression:expr, $msg:expr) => (
+        if ! $expression {
+            let error = ::ffi::ERR_get_error();
+            let bytes = ::std::ffi::CStr::from_ptr(::ffi::ERR_error_string(error, ::std::ptr::null_mut())).to_bytes();
+            panic!("{}: {}", $msg, ::std::str::from_utf8(bytes).unwrap());
+        }
+    )
+}
+
+impl<'a> Signer<'a> {
+    /// Create and initialize a new Signer
+    ///
+    /// The digest type specified will determine the final form  of the
+    /// signature returned by `finalize`.
+    ///
+    /// # Panics
+    ///
+    /// If the provided pkey does not contain a private key.
+    pub fn new(digest_type: hash::Type, pkey: &'a PKey) -> Signer {
+        ffi::init();
+
+        let ctx = unsafe {
+            let r = ffi::EVP_MD_CTX_create();
+            openssl_assert!(!r.is_null(), "EVP_MD_CTX_create failed");
+            r
+        };
+
+        let md = digest_type.evp_md();
+
+        let signer = Signer {
+            ctx: ctx,
+            pkey: pkey,
+        };
+
+        unsafe {
+            let r =
+                ffi::EVP_DigestSignInit(ctx, ptr::null_mut(), md, ptr::null(), signer.pkey.get_handle());
+            openssl_assert!(r == 1, "EVP_DigestSignInitFailed");
+        }
+
+        signer
+    }
+
+    /// Feed bytes to be added to the signature calculation
+    #[inline]
+    fn update(&mut self, data: &[u8]) {
+        unsafe {
+            let r = ffi::EVP_DigestUpdate(self.ctx, data.as_ptr(), data.len() as u32);
```

Comment:
I suppose we could map things into multiple underlying update calls, although I sure hope nobody is trying pass chunks longer than can be represented with 32 bits of length (~4GB) to update.  That is not generally a good way to do things (maybe OK with memory mapped IO?).